Allow referencing a non-PodSpec-able application in Custom Projection (#100)
k8s-service-bindings/spec: README.md

Restricting service binding to resources within the same namespace is strongly **RECOMMENDED**.  Cross-namespace service binding **SHOULD** be secured appropriately by the implementor to prevent attacks like privilege escalation and secret enumeration.

> A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration to a `PodSpec`-able resource.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.  Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.

Reviewer: Why include "services" here? I don't see any further details about referencing "services" in other ways under the custom projections section.

Also the word "kinds of" reads as this is something about Kubernetes resource kinds. Probably the sentence can be rephrased something like this:
`Extensions and implementations **MAY** allow additional mechanisms for referencing applications`
or
`Extensions and implementations **MAY** allow additional ways for referencing applications`

Reviewer 2: > Why include "services" here? I don't see any further details about referencing "services" in other ways under the custom projections section.

See https://github.com/k8s-service-bindings/spec#binding-secret-generation-strategies

> Also the word "kinds of" reads as this is something about Kubernetes resource kinds.

Correct. Other Kubernetes resource types (kinds) may be referenced. We are **not** encuraging other ways to define the reference.

Reviewer: Thanks for the clarification!

Author: @baijum Are you good with the following or you think it needs to be more clear?
> Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.

Reviewer: I am good with these changes!

> A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration to a `PodSpec`-able resource.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  The application reference by-name **MAY** be extended through [Custom Projection][cp] to reference a non-`PodSpec`-able resource in addition to a `PodSpec`-able resource.  A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.

Reviewer 2: The 'Custom Projection' is an example of an extension/implementation adding additional non-portable behavior. We  should make sure that it is viewed as an exception, but not the only exception.

What if we changed the wording to something like:
> Extensions and implementations **MAY** allow additional types of applications and services to be referenced.

Author: Good point. I updated it

Reconciler B **MUST** set the `ServiceBindingProjection`'s `Ready` condition according to the rules set in [Ready Condition Status](#ready-condition-status). Reconciler A **MUST** reflect `ServiceBindingProjection`'s `Ready` condition with an additional `ProjectionReady` condition on `ServiceBinding`.  The `Ready` condition of `ServiceBinding` **MUST NOT** be `True` unless `ProjectionReady` is `True`.

> An implementation of the Custom Projection extension **MAY** allow defining a `.spec.application` referencing a non-`PodSpec`-able resource in addition to the mechanisms to define an application reference described in the [Service Binding][sb] section.

Reviewer: This change is allowed in the `ServiceBinding` resource and not the `ServiceBindingProjection` resource, right?
Please clarify that with an additional sentence.

Reviewer 2: The primary desire for the `ServiceBindingProjection` is to have a way for non-PodSpecable resources to be the target of the ServiceBinding's application reference. Strictly speaking, the PodSpecable requirement doesn't exist on the ServiceBindingProjection resource, so there is nothing to relax, but I agree that a clarification is worthwhile.

Reviewer 2: This requirement applies to the ServiceBinding resource. Saying "An implementation of the Custom Projection extension" is redundant because it's in the Custom Projection section, it could only ever apply to the Custom Projection implementation.

```suggestion
A `ServiceBinding` with a custom projection **MUST** allow defining a `.spec.application` referencing a non-`PodSpec`-able resource in addition to the mechanisms to define an application reference described in the [Service Binding][sb] section.
```

Reviewer: This change avoids the confusion!

Author: Thanks @baijum and @scothis. I committed the suggestions
